File: apps/api/src/controllers/application.controller.ts
import { Body, Controller, Get, Param, Post } from "@nestjs/common";
import { Application } from "@prisma/client";
import { LogLines } from "engine";

import { ApiResponse } from "src/dtos/api-response.dto";
import { ApplicationResponse, CreateApplication } from "src/dtos/application.dto";
import { ApplicationService } from "src/services/application.service";

@Controller('applications')
export class ApplicationController {

    constructor(private service: ApplicationService) { }

    @Get()
    async findAll(): Promise<ApiResponse<ApplicationResponse[]>> {
        try {
            return {
                statusCode: 200,
                data: await this.service.findAll()
            }
        } catch (error) {
            return {
                statusCode: 500,
                error: error
            }
        }

    }

    @Get(':id')
    async findById(@Param('id') id: string): Promise<ApiResponse<ApplicationResponse>> {
        try {
            return {
                statusCode: 200,
                data: await this.service.findById(id)
            }
        } catch (error) {
            return {
                statusCode: 500,
                error
            }
        }
    }

    @Get(':name/logs')
    async findLogs(@Param('name') name: string): Promise<ApiResponse<LogLines[]>> {
        try {
            return {
                statusCode: 200,
                data: await this.service.findLogs(name)
            }
        } catch (error) {
            return {
                statusCode: 500,
                error
            }
        }

    }

    @Post()
    async create(@Body() data: CreateApplication): Promise<ApiResponse<Application>> {
        try {
            return {
                statusCode: 200,
                data: await this.service.create(data)
            }
        } catch (error) {
            return {
                statusCode: 500,
                error
            }
        }
    }
}

File: backend/src/main/java/com/esmo/empaas/controllers/ApplicationController.java
package com.esmo.empaas.controllers;

import java.util.List;
import java.util.Map;

import org.modelmapper.ModelMapper;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.esmo.empaas.dtos.ApplicationDto;
import com.esmo.empaas.entities.ApplicationEntity;
import com.esmo.empaas.services.ApplicationService;

import io.kubernetes.client.openapi.models.V1Deployment;

@RestController
@RequestMapping("/applications")
public class ApplicationController {

	private final ApplicationService service;
	private final ModelMapper mapper;

	public ApplicationController(ApplicationService service, ModelMapper mapper) {
		this.service = service;
		this.mapper = mapper;
	}

	@GetMapping()
	public ResponseEntity<List<ApplicationDto>> findAll() {
		return ResponseEntity.ok(service.findAll().stream().map(e -> mapper.map(e, ApplicationDto.class)).toList());
	}

	@GetMapping("/{id}")
	public ResponseEntity<ApplicationDto> findById(@PathVariable String id) {
		return ResponseEntity.ok(mapper.map(service.findById(id), ApplicationDto.class));
	}

	@GetMapping("/{id}/logs")
	public ResponseEntity<Map<String, String>> findLogs(@PathVariable String id) {
		return ResponseEntity.ok(service.findLogs(id));
	}

	@PostMapping("/{id}/deploy")
	public ResponseEntity<V1Deployment> deploy(@PathVariable String id) {
		return ResponseEntity.ok(service.deploy(id));

//		return ResponseEntity.ok("Application deployed");
	}

	@PostMapping()
	public ResponseEntity<ApplicationDto> create(@RequestBody ApplicationDto data) {
		return ResponseEntity
				.ok(mapper.map(service.create(mapper.map(data, ApplicationEntity.class)), ApplicationDto.class));
	}

	@PutMapping("/update")
	public ResponseEntity<ApplicationDto> update(@RequestParam String id, @RequestBody ApplicationDto data) {
		return ResponseEntity
				.ok(mapper.map(service.update(id, mapper.map(data, ApplicationEntity.class)), ApplicationDto.class));
	}

	@DeleteMapping("/delete")
	public ResponseEntity<Void> delete(@RequestParam String id) {
		service.delete(id);

		return ResponseEntity.ok().build();
	}
}
